src/data_handling/ApogeePredictor.cpp

```diff
@@ -0,0 +1,125 @@
+    if (apogee)
+    {
+        #ifdef DEBUG
+        Serial.println("ApogeePredictor: Data point ignored because already at apogee");
+        #endif
+        return false;
+    }
+
+    // Calculate the magnitude of the acceleration squared
+    // May only want to use the z acceleration as y and x could move around at apogee
+    float aclMagSq = xac.data * xac.data + yac.data * yac.data + zac.data * zac.data;
+    float accel = direction == 0 ? xac.data : direction == 1 ? yac.data : zac.data;
+
+    // Take the average of the timestamps
+    // Ideally these should all be the same
+    uint32_t time_ms = (xac.timestamp_ms + yac.timestamp_ms + zac.timestamp_ms) / 3;
+
+    // Making sure the new time is greater than the last time
+    // Potentially could be a problem if the timestamps are not in sequenital order and we don't care
+    if (time_ms < AclMagSqWindow_ms2.getFromHead(0).timestamp_ms)
+    {
+        #ifdef DEBUG
+        Serial.println("LaunchPredictor: Data point ignored because of time is earlier than head");
+        Serial.printf("Incoming time: %d\n", time_ms);
+        Serial.printf("Head time: %d\n", AclMagSqWindow_ms2.getFromHead(0).timestamp_ms);
+        #endif
+        return false;
+    }
+
+    // Make sure we are near the window interval +- 10%
+    // Again if we want to ignore doing time windowing we can remove this
+    // Maybe get a system so that time can't change by that must be does not have to be linear
+    uint32_t time_diff = time_ms - AclMagSqWindow_ms2.getFromHead(0).timestamp_ms;
+
+    if (time_diff < windowInterval_ms - tenPercentWindowInterval_ms || time_diff > windowInterval_ms + tenPercentWindowInterval_ms)
+    {
+        #ifdef DEBUG
+        Serial.println("LaunchPredictor: Data point ignored because of time difference");
+        Serial.printf("Time diff: %d\n", time_diff);
+        Serial.printf("Window interval: %d\n", windowInterval_ms);
+        Serial.printf("Incoming time: %d\n", time_ms);
+        Serial.printf("Head time: %d\n", AclMagSqWindow_ms2.getFromHead(0).timestamp_ms);
+        #endif
+        return false;
+    }
+
+    // Add the new data point to the window
+    AclMagSqWindow_ms2.push(DataPoint(time_ms, aclMagSq));
+    velocity = VelocityWindow_ms2.getMedian().data + accel * (windowInterval_ms / 1000);
+    VelocityWindow_ms2.push(DataPoint(time_ms, velocity));
+    // Check that the window is full
+    if (!AclMagSqWindow_ms2.isFull())
+    {
+        return true;
+    }
+
+    // Maybe should be if the acceleration is 0 but more testing needs to be done
+    // or velocity is 0
+    if (VelocityWindow_ms2.getMedian().data == 0) {
+        #ifdef DEBUG
+        Serial.println("ApogeePredictor: Velocity is 0");
+        Serial.println("Current time: %d", time_ms);
+        Serial.println("Current accel: %f", accel);
+        #endif
+    }
+
+    // Check if the median is equal to 0, acceleration is 0
+    if (AclMagSqWindow_ms2.getMedian().data == 0)
```

Comment:
The median will never be exactly zero. Maybe?

Reply:
Yes that is correct I am still trying to figure out how to find direction with the accelerations

Reply:
It will probably end up being something close enough to zero

Reply:
How close to zero does it typically get? Maybe += .1 

`if abs(median) < .1 then apogee`

Reply:
At least of the AclMagSq it will never be negative as its squared so yes < .1 should work; however, I want to also have an velocityWindow circular array at that could be negative so I may make some new logic for both

Reply:
I'd go for maybe a bit of a wider window for that, you get a lot of wacky forces acting upon the rocket at apogee that could shift that number slightly outside of where it should be even at apogee. Technically we should be sampling at a high enough rate where this will be fine but still make sure to double check this with tests.

Reply:
Alright after reviewing flight data from a year ago, it does seem to drop below 0.1 but only somewhat briefly, so make sure you have a high polling rate on that sensor when using it.